src/main/scala/org/broadinstitute/hail/keytable/KeyTable.scala

```diff
@@ -0,0 +1,397 @@
+
+  require(fieldNames.areDistinct())
+
+  def signature = keySignature.merge(valueSignature)._1
+
+  def fields = signature.fields
+
+  def keySchema = keySignature.schema
+
+  def valueSchema = valueSignature.schema
+
+  def schema = signature.schema
+
+  def keyNames = keySignature.fields.map(_.name).toArray
+
+  def valueNames = valueSignature.fields.map(_.name).toArray
+
+  def fieldNames = keyNames ++ valueNames
+
+  def nRows = rdd.count()
+
+  def nFields = fields.length
+
+  def nKeys = keySignature.size
+
+  def nValues = valueSignature.size
+
+  def same(other: KeyTable): Boolean = {
+    if (fields.toSet != other.fields.toSet) {
+      println(s"signature: this=${ schema } other=${ other.schema }")
+      false
+    } else if (keyNames.toSet != other.keyNames.toSet) {
+      println(s"keyNames: this=${ keyNames.mkString(",") } other=${ other.keyNames.mkString(",") }")
+      false
+    } else {
+      val thisFieldNames = valueNames
+      val otherFieldNames = other.valueNames
+
+      rdd.groupByKey().fullOuterJoin(other.rdd.groupByKey()).forall { case (k, (v1, v2)) =>
+        (v1, v2) match {
+          case (None, None) => true
+          case (Some(x), Some(y)) =>
+            val r1 = x.map(r => thisFieldNames.zip(r.asInstanceOf[Row].toSeq).toMap).toSet
+            val r2 = y.map(r => otherFieldNames.zip(r.asInstanceOf[Row].toSeq).toMap).toSet
+            val res = r1 == r2
+            if (!res)
+              println(s"k=$k r1=${ r1.mkString(",") } r2=${ r2.mkString(",") }")
+            res
+          case _ =>
+            println(s"k=$k v1=$v1 v2=$v2")
+            false
+        }
+      }
+    }
+  }
+
+  def mapAnnotations[T](f: (Annotation) => T)(implicit tct: ClassTag[T]): RDD[T] =
+    KeyTable.toSingleRDD(rdd, nKeys, nValues).map(a => f(a))
+
+  def mapAnnotations[T](f: (Annotation, Annotation) => T)(implicit tct: ClassTag[T]): RDD[T] =
+    rdd.map { case (k, v) => f(k, v) }
+
+  def query(code: String): (BaseType, (Annotation, Annotation) => Option[Any]) = {
+    val ec = EvalContext(fields.map(f => (f.name, f.`type`)): _*)
+    val nKeysLocal = nKeys
+    val nValuesLocal = nValues
+
+    val (t, f) = Parser.parse(code, ec)
+
+    val f2: (Annotation, Annotation) => Option[Any] = {
+      case (k, v) =>
+        KeyTable.setEvalContext(ec, k, v, nKeysLocal, nValuesLocal)
+        f()
+    }
+
+    (t, f2)
+  }
+
+  def querySingle(code: String): (BaseType, Querier) = {
+    val ec = EvalContext(fields.map(f => (f.name, f.`type`)): _*)
+    val nFieldsLocal = nFields
+
+    val (t, f) = Parser.parse(code, ec)
+
+    val f2: (Annotation) => Option[Any] = { a =>
+      KeyTable.setEvalContext(ec, a, nFieldsLocal)
+      f()
+    }
+
+    (t, f2)
+  }
+
+  def annotate(cond: String, keyNameString: String): KeyTable = {
+    val ec = EvalContext(fields.map(fd => (fd.name, fd.`type`)): _*)
+
+    val (parseTypes, fns) =
+      if (cond != null)
```

Comment:
When is cond going to be null?  This looks fishy.  Either the `else` branch is redundant, or we're somehow encoding missingness without options here.